From 39fee88ca03af04ce41a9a872567f66c70862c91 Mon Sep 17 00:00:00 2001
From: Hendrik Leppkes <h.leppkes@gmail.com>
Date: Mon, 28 Mar 2011 22:39:52 +0200
Subject: [PATCH 3/5] Added bd_get_clip_infos

This function allows for querying information directly related to the
clips inside a title.
---
 libbluray.def          |  1 +
 src/libbluray/bluray.c | 17 +++++++++++++++++
 src/libbluray/bluray.h | 13 +++++++++++++
 3 files changed, 31 insertions(+)

diff --git a/src/libbluray/bluray.c b/src/libbluray/bluray.c
index d19872f..5fe5c53 100644
--- a/src/libbluray/bluray.c
+++ b/src/libbluray/bluray.c
@@ -3292,3 +3292,20 @@ void bd_free_clpi(struct clpi_cl *cl)
 {
     clpi_free(cl);
 }
+
+int bd_get_clip_infos(BLURAY *bd, int clip, uint64_t *clip_start_time, uint64_t *stream_start_time, uint64_t *pos, uint64_t *duration)
+{
+    if (bd && bd->title && bd->title->clip_list.count > clip) {
+      if (clip_start_time)
+        *clip_start_time = (uint64_t)bd->title->clip_list.clip[clip].start_time << 1;
+      if (stream_start_time)
+        *stream_start_time = (uint64_t)bd->title->clip_list.clip[clip].in_time << 1;
+      if (pos)
+        *pos = (uint64_t)bd->title->clip_list.clip[clip].pos * 192;
+      if (duration)
+        *duration = (uint64_t)bd->title->clip_list.clip[clip].duration << 1;
+
+      return 1;
+    }
+    return 0;
+}
diff --git a/src/libbluray/bluray.h b/src/libbluray/bluray.h
index 828a74c..22cb7bc 100644
--- a/src/libbluray/bluray.h
+++ b/src/libbluray/bluray.h
@@ -861,6 +861,19 @@ struct clpi_cl *bd_read_clpi(const char *clpi_file);
  */
 void bd_free_clpi(struct clpi_cl *cl);
 
+/**
+ *
+ * Get information about the clip
+ *
+ * @param bd  BLURAY object
+ * @param clip clip index
+ * @param clip_start_time start of the clip (in the total title) (in 90khz)
+ * @param stream_start_time first pts in the clip (in 90khz)
+ * @param byte position of the clip (absolute)
+ * @param duration duration of the clip (in 90khz)
+ */
+int bd_get_clip_infos(BLURAY *bd, int clip, uint64_t *clip_start_time, uint64_t *stream_start_time, uint64_t *pos, uint64_t *duration);
+
 #ifdef __cplusplus
 };
 #endif
-- 
1.8.3.1

